lib/seadsa/SeaDsaAliasAnalysis.cc

```diff
@@ -1,41 +1,207 @@
 #include "llvm/IR/Function.h"
 #include "llvm/IR/PassManager.h"
 #include "llvm/IR/Type.h"
-
+#include "seadsa/AllocWrapInfo.hh"
 #include "seadsa/InitializePasses.hh"
 #include "seadsa/support/Debug.h"
+#include "seadsa/Global.hh"
+#include "seadsa/Graph.hh"
+
 
 #define DEBUG_TYPE "sea-aa"
 using namespace llvm;
 using namespace seadsa;
 namespace dsa = seadsa;
 
 namespace seadsa {
+  
 SeaDsaAAResult::SeaDsaAAResult(
-    std::function<const TargetLibraryInfo &(Function &F)> GetTLI)
-    : GetTLI(std::move(GetTLI)) {}
+    TargetLibraryInfoWrapperPass &tliWrapper,
+    AllocWrapInfo &awi)
+  : m_tliWrapper(tliWrapper), m_awi(awi), m_fac(nullptr), m_dsa(nullptr)  {}
+  
 SeaDsaAAResult::SeaDsaAAResult(SeaDsaAAResult &&RHS)
-    : AAResultBase(std::move(RHS)), GetTLI(std::move(RHS.GetTLI)) {}
+  : AAResultBase(std::move(RHS)), m_tliWrapper(RHS.m_tliWrapper), m_awi(RHS.m_awi),
+    m_fac(std::move(RHS.m_fac)), m_dsa(std::move(RHS.m_dsa)) {}
+  
 SeaDsaAAResult::~SeaDsaAAResult() = default;
 
+static Module* getModuleFromQuery(Value *ValA, Value *ValB) {
+  Function *MaybeFnA = const_cast<Function *>(llvm::cflaa::parentFunctionOfValue(ValA));
+  Function *MaybeFnB = const_cast<Function *>(llvm::cflaa::parentFunctionOfValue(ValB));
+  if (!MaybeFnA && !MaybeFnB) {
+    // The only times this is known to happen are when globals + InlineAsm are
+    // involved
+    DOG(llvm::errs()
+	<< "SeaDsaAA: could not extract parent function information.\n");
+    return nullptr;
+  }
+  Module *M = nullptr;
+  if (MaybeFnA) {
+    M = MaybeFnA->getParent();
+  }
+  if (MaybeFnB) {
+    if (M != MaybeFnA->getParent()) {
+      DOG(llvm::errs()
+	  << "SeaDsaAA: cannot handle functions in different modules.\n");
+      return nullptr;
+    }
+  }
+  return M;
+}
+
+static uint64_t storageSize(const Type *t, const DataLayout &dl) {
+  return dl.getTypeStoreSize(const_cast<Type *>(t));
+}
+
+static Optional<uint64_t> sizeOf(const Graph::Set &types, const DataLayout &dl) {
+  if (types.isEmpty()) {
+    return 0;
+  } else {
+    uint64_t sz = storageSize(*(types.begin()), dl);
+    if (types.isSingleton()) {
+      return sz;
+    } else {
+      auto it = types.begin();
+      ++it;
+      if (std::all_of(it, types.end(), [dl, sz](const Type *t) {
+	      return (storageSize(t, dl) == sz);
+          })) {
+        return sz;
+      } else {
+        return None;
+      }
+    }
+  }
+}
+  
+static bool mayAlias(const Cell &c1, const Cell &c2, const DataLayout &dl) {
+  auto maybeUnsafe = [](const Node *n) {
+		       return n->isIntToPtr() || n->isPtrToInt() ||
+			 n->isIncomplete() || n->isUnknown();
+		     };
+  
+  if (c1.isNull() || c2.isNull()) {
+    return true;
+  }
+  
+  const Node *n1 = c1.getNode();
+  const Node *n2 = c2.getNode();
+
+  if (maybeUnsafe(n1) || maybeUnsafe(n2)) {
+    return true;
+  }
+
+  if (n1 != n2) {
+    // different nodes cannot alias
+    return false;
+  }
+
+  if (n1->isOffsetCollapsed() || n1->isArray()) {
```

Comment:
pointers into an array at different offsets do not alias. Think of array of structs with fields `f1` and `f2`. Pointer to `f1` does not alias with pointer to `f2`